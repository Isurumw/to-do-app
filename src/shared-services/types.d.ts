declare interface IStorage {
  get(key: string): Promise<string | undefined>;
  getObject<P>(key: string): Promise<P | undefined>;
  set(key: string, value: string): Promise<void>;
  remove(key: string): void;
}

declare interface INetworkManager {
  request(
    route: string,
    method: 'POST' | 'GET' | 'PUT' | 'DELETE' | 'PATCH',
    variables?: object | null,
  ): any;
}
