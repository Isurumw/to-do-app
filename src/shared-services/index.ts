import Config from 'react-native-config';
import AsyncStorage from '@react-native-community/async-storage';

import _NetworkManager from './network-services/network-manager';
import _StorageService from './storage-service/storage-service';

const StorageService = new _StorageService(AsyncStorage);
const NetworkManager = new _NetworkManager(
  Config.AUTH_BASE,
  Config.AUTH_KEY,
  StorageService,
  fetch,
);
const FirestoreManager = new _NetworkManager(
  Config.FIRESTORE_BASE,
  Config.AUTH_KEY,
  StorageService,
  fetch,
);

export {NetworkManager, FirestoreManager, StorageService};
