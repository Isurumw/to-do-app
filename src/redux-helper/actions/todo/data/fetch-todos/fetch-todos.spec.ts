import {FirestoreManager} from 'shared-services';
import fetchTodos from './fetch-todos';

jest.mock('shared-services/network-services/network-manager');
const mockedFirestoreManager =
  FirestoreManager as unknown as jest.Mocked<INetworkManager>;

const mockResponse = {
  documents: [
    {
      name: 'projects/todo-8013f/databases/(default)/documents/todos/4tv4cXgCJYqalcTNaUCT',
      fields: {
        description: {
          stringValue: 'Decide what medium works best for you',
        },
        category: {
          stringValue: 'work',
        },
      },
      createTime: '2021-09-28T06:52:27.470415Z',
    },
    {
      name: 'projects/todo-8013f/databases/(default)/documents/todos/AQq016UeVdVxkXGjBMve',
      fields: {
        description: {
          stringValue: 'Apply for uLethbridge awards and scholarships',
        },
        category: {
          stringValue: 'school',
        },
      },
      createTime: '2021-09-28T06:40:35.637640Z',
    },
    {
      name: 'projects/todo-8013f/databases/(default)/documents/todos/DOsnkr0lMwVzhU5D77ZU',
      fields: {
        description: {
          stringValue: 'Write down your goals and priorities',
        },
        category: {
          stringValue: 'personal',
        },
      },
      createTime: '2021-09-28T06:44:49.564824Z',
    },
  ],
};

const mockResult = [
  {
    name: 'projects/todo-8013f/databases/(default)/documents/todos/4tv4cXgCJYqalcTNaUCT',
    fields: {
      description: {
        stringValue: 'Decide what medium works best for you',
      },
      category: {
        stringValue: 'work',
      },
    },
    createTime: '2021-09-28T06:52:27.470415Z',
  },
  {
    name: 'projects/todo-8013f/databases/(default)/documents/todos/AQq016UeVdVxkXGjBMve',
    fields: {
      description: {
        stringValue: 'Apply for uLethbridge awards and scholarships',
      },
      category: {
        stringValue: 'school',
      },
    },
    createTime: '2021-09-28T06:40:35.637640Z',
  },
  {
    name: 'projects/todo-8013f/databases/(default)/documents/todos/DOsnkr0lMwVzhU5D77ZU',
    fields: {
      description: {
        stringValue: 'Write down your goals and priorities',
      },
      category: {
        stringValue: 'personal',
      },
    },
    createTime: '2021-09-28T06:44:49.564824Z',
  },
];

describe('check the behaviour of the fetch-todo natwork service', () => {
  afterEach(() => {
    mockedFirestoreManager.request.mockClear();
  });

  it('fetch todos successful', async () => {
    mockedFirestoreManager.request.mockImplementation(() => mockResponse);

    const response = await fetchTodos();

    expect(mockedFirestoreManager.request).toBeCalledTimes(1);
    expect(response).toEqual(mockResult);
  });

  it('fetch todos failed with an error', async () => {
    mockedFirestoreManager.request.mockImplementation(() => {
      throw new Error('Some API Error');
    });

    expect(fetchTodos()).rejects.toThrowErrorMatchingSnapshot();
  });
});
