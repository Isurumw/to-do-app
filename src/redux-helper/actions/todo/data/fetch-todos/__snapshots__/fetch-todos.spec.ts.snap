// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check the behaviour of the fetch-todo natwork service fetch todos failed with an error 1`] = `"Some API Error"`;
